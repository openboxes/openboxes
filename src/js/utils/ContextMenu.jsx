import React from 'react';

import PropTypes from 'prop-types';
import { RiMoreLine } from 'react-icons/ri';
import { Link } from 'react-router-dom';
import { Popover } from 'react-tiny-popover';

import actionItemType from 'consts/actionItemType';
import useContextMenu from 'hooks/useContextMenu';
import Translate from 'utils/Translate';

import './utils.scss';

const ContextMenu = ({
  positions, actions, dropdownClasses, popoverClasses, id,
}) => {
  const {
    getActionItemType,
    buildLink,
    popoverRef,
    isPopoverOpen,
    setIsPopoverOpen,
  } = useContextMenu();

  return (
    <Popover
      data-testid="action-dots-component"
      isOpen={isPopoverOpen}
      positions={positions}
      padding={10}
      onClickOutside={() => setIsPopoverOpen(false)}
      ref={popoverRef}
      content={() => (
        actions && (
        <div
          className={`${popoverClasses} context-menu padding-8`}
          data-testid="dropdown-menu"
        >
          {actions.map((action) => {
            const itemClasses = `${dropdownClasses} d-flex align-items-center gap-8 dropdown-item ${action.variant === 'danger' ? 'font-red-ob' : ''}`;
            const itemValue = (
              <>
                {action.leftIcon && action.leftIcon}
                {action.label
                  && <Translate id={action.label} defaultMessage={action.defaultLabel} />}
              </>
            );
            const elementType = getActionItemType(action);
            const link = elementType === actionItemType.LINK || actionItemType.REACT_LINK
              ? buildLink(action, id)
              : '';

            return (
              <React.Fragment key={action.label}>
                {elementType === actionItemType.BUTTON && (
                <button
                  type="button"
                  onClick={() => {
                    action.onClick(id);
                    setIsPopoverOpen(false);
                  }}
                  className={itemClasses}
                >
                  {itemValue}
                </button>
                )}
                {elementType === actionItemType.LINK && (
                <a href={link} className={itemClasses}>
                  {itemValue}
                </a>
                )}
                {elementType === actionItemType.REACT_LINK && (
                <Link to={link} className={itemClasses}>
                  {itemValue}
                </Link>
                )}
              </React.Fragment>
            );
          })}
        </div>
        )
      )}
    >
      <button
        data-testid="dropdown-toggle"
        className="action-dots dropdown-toggle d-flex align-items-center justify-content-center"
        type="button"
        onClick={() => setIsPopoverOpen((isOpen) => !isOpen)}
      >
        <RiMoreLine />
      </button>
    </Popover>
  );
};

export default ContextMenu;

ContextMenu.propTypes = {
  actions: PropTypes.arrayOf(PropTypes.shape({
    leftIcon: PropTypes.element.isRequired,
    label: PropTypes.string.isRequired,
    defaultLabel: PropTypes.string.isRequired,
    href: PropTypes.string,
    reactLink: PropTypes.bool,
    variant: PropTypes.string,
    onClick: PropTypes.func,
  })).isRequired,
  positions: PropTypes.arrayOf(PropTypes.string),
  id: PropTypes.string.isRequired,
  dropdownClasses: PropTypes.string,
  popoverClasses: PropTypes.string,
};

ContextMenu.defaultProps = {
  positions: ['top', 'right', 'left', 'bottom'],
  dropdownClasses: '',
  popoverClasses: '',
};
