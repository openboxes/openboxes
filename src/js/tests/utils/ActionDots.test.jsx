import React from 'react';

import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import ContextMenu from 'utils/ContextMenu';

import '@testing-library/jest-dom';

let actionDots;
let actions;

// jest doesn't have access to bootstrap styles, so I included it here.
const applyStyles = () => {
  const style = document.createElement('style');
  style.innerHTML = ` 
           .dropdown-menu { 
              display: none;
           }
           .show {
              display: block !important;
           }`;
  document.body.appendChild(style);
};

describe('action dots', () => {
  beforeEach(() => {
    applyStyles();
    actions = [{
      variant: 'danger',
      label: 'testLabel',
      defaultLabel: 'defaultLabelTest',
      leftIcon: <>leftIconTest</>,
    }];
    actionDots = (
      <ContextMenu
        id="1"
        actions={actions}
      />
    );
  });

  it('should match snapshot', () => {
    const renderedActionDots = renderer.create(actionDots);
    expect(renderedActionDots.toJSON())
      .toMatchSnapshot();
  });

  it('should render component correctly', () => {
    const renderedActionDots = renderer.create(actionDots);
    expect(renderedActionDots.root.findByProps({ 'data-testid': 'action-dots-component' }))
      .toBeTruthy();
  });

  it('should have button after clicking it', () => {
    const testId = 'dropdown-toggle';
    render(actionDots);
    expect(screen.findByTestId(testId))
      .toBeTruthy();
    fireEvent.click(screen.getByTestId(testId));
    expect(screen.findByTestId(testId))
      .toBeTruthy();
  });
});
