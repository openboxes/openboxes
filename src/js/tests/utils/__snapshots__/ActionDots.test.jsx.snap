// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action dots should match snapshot 1`] = `
<button
  className="action-dots dropdown-toggle d-flex align-items-center justify-content-center"
  data-testid="dropdown-toggle"
  onClick={[Function]}
  type="button"
>
  <svg
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      {
        "color": undefined,
      }
    }
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g>
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M4.5 10.5c-.825 0-1.5.675-1.5 1.5s.675 1.5 1.5 1.5S6 12.825 6 12s-.675-1.5-1.5-1.5zm15 0c-.825 0-1.5.675-1.5 1.5s.675 1.5 1.5 1.5S21 12.825 21 12s-.675-1.5-1.5-1.5zm-7.5 0c-.825 0-1.5.675-1.5 1.5s.675 1.5 1.5 1.5 1.5-.675 1.5-1.5-.675-1.5-1.5-1.5z"
      />
    </g>
  </svg>
</button>
`;
